independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/RunnerClassLoader.java

```diff
@@ -64,26 +65,39 @@ public final class RunnerClassLoader extends ClassLoader {
         pooledData = null;
     }
 
+    public void setMainThreadId(long mainThreadId) {
+        this.mainThreadId = mainThreadId;
+    }
+
     @Override
     public Class<?> loadClass(String name) throws ClassNotFoundException {
         return loadClass(name, false);
     }
 
     private ByteBuffer acquirePooledData() {
-        ByteBuffer pooledData = null;
-        if (!postBootPhase) {
-            pooledData = this.pooledData;
-            // this will make sure that the pooledData is not used again in any recursive call to loadClass
-            this.pooledData = null;
+        var main = mainThreadId;
+        if (main < 0) {
+            return null;
+        }
+        // how many chances tearing on mainThreadId could match this thread id? 64-bit archs shouldn't make it happen!
+        if (Thread.currentThread().getId() != main) {
+            return null;
         }
+        final var pooledData = this.pooledData;
```

Comment:
This can happen only if the main thread die and its id is reused by a new thread which is calling here; but this is something we expect to happen? @geoand @dmlloyd

Reply:
The thread ID is a 64-bit value, sequentially counted (IIRC), so I would have expected the answer to be "no". But this is out of date anyway, so I'll just resolve it. :)